design/control-data-plane-separation/design.md

```diff
@@ -23,377 +23,118 @@ architecture does not allow the control plane and data plane to scale independen
   use the DNS name of the Service to connect to the control plane.
-- _Data Plane DaemonSet/Deployment_: A user can deploy the data plane as either a DaemonSet or Deployment. The data
+- _Data Plane DaemonSet/Deployment_: The data plane can be deployed as either a DaemonSet or Deployment. The data
   plane contains a single container running both the agent and nginx processes. The agent will download the
   configuration from the control plane over a streaming RPC.
-- _NGINX Service_: Exposes nginx via a Kubernetes Service of type `LoadBalancer .`This is the entry point for the
-  customer’s traffic. Initially, this Service will only expose ports 80 and 443. In the future, if we add support for
-  additional listener ports, this Service will expose all the listener ports. Note that this Service should not expose
-  any of the agent’s ports.
+- _NGINX Service_: Exposes nginx via a Kubernetes Service of type `LoadBalancer .` This is the entry point for the
+  customer’s traffic. Note that this Service should not expose any of the agent’s ports.
+
+#### Further Requirements and Implementation Details
+
+- Both deployments should have read only filesystems.
+- Both deployments should have the minimal permissions required to perform their functions.
+- The nginx deployment should be configurable via the helm chart.
+  - Downside of this is that these options will apply to all nginx instances.
+  - We could introduce a CRD, but where would it attach? We already have NginxProxy which controls dynamic data plane configuration, and this may eventually attach to the Gateway instead of just the GatewayClass. Would a Deployment configuration fit in there, and would it be dynamic? That would require us to completely redeploy nginx if a user changes those settings.
+  - We could start with the helm chart option, and rely on user feedback to see if we need to get more granular.
+  - This could also involve creating a ConfigMap that the control plane consumes on startup and contains all nginx Deployment/Daemonset configuration, including NGINX Plus usage configuration.
+- nginx Service should have configurable labels and annotations via the GatewayInfrastructure field in the Gateway resource.
+- Control plane creates the nginx deployment and service when a Gateway resource is created, in the same namespace as the Gateway resource. When the Gateway is deleted, the control plane deletes nginx deployment and service.
+- Control plane should label the nginx service and deployment with something related to the name of the Gateway so it can easily be linked.
+- Liveness/Readiness probes:
+  - Control plane probe currently waits until we configure nginx. Going forward, this probe should just be when the control plane is ready to configure.
+  - nginx probe...can agent expose any type of health endpoint? This is not yet clear.
+- Control plane should not restart data plane pods if they are unhealthy.
```

Comment:
So we are relying on Kubernetes to restart the data plane pods? That means we will need the agent to exit on unrecoverable failures.

Reply:
This came out of the discussion we had in person. I think the intention was to leave it in the user's hands, not necessarily automatic restart by kubernetes. Although if a probe exists that marks it as unready, the k8s might just do that anyway. I can clarify.